x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/LoadAuthorizedIndicesTimeChecker.java

```diff
@@ -0,0 +1,187 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authz;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.settings.SettingsException;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xpack.core.security.authz.AuthorizationEngine;
+
+import java.util.Collection;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+/**
+ * Checks the time spent evaluating authorized indices for a request.
+ * Has configurable logging levels based on execution time.
+ */
+class LoadAuthorizedIndicesTimeChecker implements Consumer<Collection<String>> {
+
+    private final Logger logger;
+    private final long startNanos;
+    private final AuthorizationEngine.RequestInfo requestInfo;
+    private final Thresholds thresholds;
+
+    LoadAuthorizedIndicesTimeChecker(Logger logger, long startNanos, AuthorizationEngine.RequestInfo requestInfo, Thresholds thresholds) {
+        this.logger = logger;
+        this.startNanos = startNanos;
+        this.requestInfo = requestInfo;
+        this.thresholds = thresholds;
+    }
+
+    @Override
+    public void accept(Collection<String> indices) {
+        done(indices);
+    }
+
+    private void done(Collection<String> indices) {
+        final long end = System.nanoTime();
+        final long millis = TimeUnit.NANOSECONDS.toMillis(end - startNanos);
+        final Level level = thresholds.getLogLevel(millis);
+        if (level == Level.WARN) {
+            logger.warn(
+                "Resolving [{}] indices for action [{}] and user [{}] took [{}ms] which is greater than the threshold of {}ms;"
+                    + " The index privileges for this user may be too complex for this cluster.",
+                indices.size(),
+                requestInfo.getAction(),
+                requestInfo.getAuthentication().getUser().principal(),
+                millis,
+                thresholds.warnThresholdMs
+            );
+        } else {
+            logger.log(
+                level,
+                "Took [{}ms] to resolve [{}] indices for action [{}] and user [{}]",
+                millis,
+                indices.size(),
+                requestInfo.getAction(),
+                requestInfo.getAuthentication().getUser().principal()
+            );
+        }
+    }
+
+    static final Setting<Boolean> LOGGING_ENABLED_SETTING = Setting.boolSetting(
+        "xpack.security.authz.timer.enabled",
+        false,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+    static final Setting<TimeValue> DEBUG_THRESHOLD_SETTING = Setting.timeSetting(
+        "xpack.security.authz.timer.threshold.debug",
```

Comment:
I changed it to `xpack.security.authz.timer.indices.*`
I don't think `load_authorized_indices` is really appropriate because that's not what the messages refer to, and it's an implementation detail, but the messages do say `"Took [{}ms] to resolve [{}] indices ..."` so `timer.indices` seems reasonable.